tests/framework/prototype.py

```diff
@@ -0,0 +1,208 @@
+'''
+Configuration Test Framework Abstractions: (1) DeltaSet (2) Mutator (3) Rule (4) RuleSet (5) CREvent
+'''
+
+# DeltaSet: Use `path` to specify the field that wants to mutate, and `candidates` is a list of candidate
+#           value for the field.
+# Example:  DeltaSet("spec.headGroupSpec.template.spec.containers.0.name", ['ray-head-1', 'ray-head-2', 'ray-head-3'])
+class DeltaSet:
+    def __init__(self, path, candidates):
+        self.path = path
+        self.candidates = candidates
+
+    def iterate(self):
+        for candidate in self.candidates:
+            yield candidate
+
+    # return value: (1) steps (2) None => this path does not exist
+    def find_path(self, cr) -> Optional[list[str]]:
+        steps = self.path.split('.')
+        return steps
+
+# Mutator: Mutator will start to mutate from `baseCR`. `deltaSets` is a list of DeltaSets, and each DeltaSet
+#          specifies a field that wants to mutate with multiple candidate values.
+# Example: "SimpleMutator"
+class Mutator:
+    def __init__(self, baseCR, deltaSets: list[DeltaSet]):
+        self.baseCR = baseCR
+        self.deltaSets = deltaSets
+    # You need to define your mutate() function by inheriting `Mutator`. It should return a new CR.
+    def mutate(self):
+        pass
+    # Apply delta to the base (custom resource).
+    def apply_delta(self, base, delta, steps):
+        root = copy.deepcopy(base)
+        curr = search_path(root, steps[:-1])
+        curr[steps[-1]] = delta
+        return root
+
+# Rule: Rule is used to check whether the actual cluster state is the same as our expectation after a CREvent.
+#       We can infer the expected state by CR YAML file, and get the actual cluster state by Kubernetes API.
+# Example: "HeadPodNameRule"
+class Rule:
+    def __init__(self):
+        pass
+    # The rule will only be checked when `trigger_condition` is true. For example, we will only check
+    # "HeadPodNameRule" when "spec.headGroupSpec" is defined in CR YAML file.
+    def trigger_condition(self, cr=None) -> bool:
+        return True
+    def assertRule(self, cr=None, namespace='default'):
+        pass
+
+# RuleSet: A set of Rule
+class RuleSet:
+    def __init__(self, rules: list[Rule]):
+        self.rules = rules
+    def checkRuleSet(self, cr, namespace):
+        for rule in self.rules:
+            if rule.trigger_condition(cr):
+                rule.assertRule(cr, namespace)
+
+# CREvent: Custom Resource Event can be mainly divided into 3 categories.
+#   (1) Add (create) CR (2) Update CR (3) Delete CR
+#
+# The member functions integrate together in `trigger()`.
+#   [Step1] exec(): Execute a command to trigger the CREvent. For example, create a CR by a
+#                  `kubectl apply` command.
+#   [Step2] wait(): Wait for the system to converge.
+#   [Step3] checkRuleSets(): When the system converges, check all registered RuleSets.
+class CREvent:
+    def __init__(self, cr, cmd, ruleSets: list[RuleSet], timeout, namespace):
```

Comment:
Is `trigger` intended as the only public method?

Reply:
Should there be a base class this inherits from?

Timeout and namespace don't seem like a thing that should be inherent to a CREvent.
(cr, cmd, and ruleSets do seem like things every CREvent has).

Reply:
Allow users to overwrite `exec()`

Reply:
```
def exec(self):
    raise NotImpl

class RayClusterAddEvent:
   def __init__(namespace)
```

Reply:
1. Updated `exec()` => `raise NotImplementedError`
2. Remove `cmd` => We can create k8s resources by both CLI tools and Python Kubernetes API.
3. `namespace` => I did not remove `namespace` from CREvent's constructor because we can set `namespace = None` when the custom resource is non-namespaced.

Reply:
TODO: Discuss `namespace`